### DISCUSSION TITLE
dubbo 2.6.8 流量录制

### DISCUSSION THREAD

**Discussion by XuelinHu** [alibaba/jvm-sandbox-repeater#223] [Q&A]
# maven 导包
```
   <dependency>
            <groupId>com.alibaba</groupId>
            <artifactId>dubbo-rpc-dubbo</artifactId>
            <version>2.6.8</version>
            <scope>provided</scope>
        </dependency>
```
# 服务消费端

-  被增强的方法  com.alibaba.dubbo.rpc.filter.ConsumerContextFilter#invoke

DubboConsumerPlugin 中增加以下增强模型
```
        EnhanceModel invoke = EnhanceModel.builder().classPattern("com.alibaba.dubbo.rpc.filter.ConsumerContextFilter")
                .methodPatterns(EnhanceModel.MethodPattern.transform("invoke"))
                .watchTypes(Event.Type.BEFORE, Event.Type.RETURN, Event.Type.THROWS)
                .build();

```
对应的，在com.alibaba.jvm.sandbox.repeater.plugin.dubbo.DubboConsumerEventListener#initInvocation方法，获取请求参数
```
        Object invoker;
        Object invocation;
invoker = event.argumentArray[0];
            invocation = event.argumentArray[1];
```

# 服务提供端 

- 被增强的方法 com.alibaba.dubbo.rpc.filter.ContextFilter#invoke

DubboProviderPlugin中增加以下增强模型
```
        EnhanceModel invoke = EnhanceModel.builder().classPattern("com.alibaba.dubbo.rpc.filter.ContextFilter")
                .methodPatterns(EnhanceModel.MethodPattern.transform("invoke"))
                .watchTypes(Event.Type.BEFORE, Event.Type.RETURN, Event.Type.THROWS)
                .build();

```

DubboProviderEventListener extends DubboConsumerEventListener ,获取参数时，**可以不加**
```
        Object invoker;
        Object invocation;
invoker = event.argumentArray[0];
            invocation = event.argumentArray[1];
```

**Comment by greatwhole** [comment]
我看dubbo的录制是依赖于从 onResponse 中获取 Result 的（推测是因为在 invoke 时的 Result 的 state 未完成，不能获取足够的信息）。
而 dubbo 2.6 没有 onResponse 的方法，不知道怎么能实现。

**Reply by XuelinHu** [reply]
增强这个方法 com.alibaba.dubbo.remoting.exchange.ResponseFuture#get(),P 这个方法被调用的地方有：
- com.alibaba.dubbo.rpc.protocol.dubbo.DubboInvoker#doInvoke
- com.alibaba.dubbo.rpc.protocol.dubbo.FutureAdapter#get()
- com.alibaba.dubbo.rpc.protocol.thrift.ThriftInvoker#doInvoke
- com.alibaba.dubbo.rpc.protocol.dubbo.ChannelWrappedInvoker#doInvoke

或者增强com.alibaba.dubbo.rpc.protocol.AbstractInvoker#doInvoke这个方法，DubboInvoker，ThriftInvoker是AbstractInvoker的子类

**Reply by greatwhole** [reply]
有增强的思路吗？现在是 invoke 的时候是 Result 未完成的 Future ，找不到入参是完成的 Result 的函数

**Reply by XuelinHu** [reply]
增强com.alibaba.dubbo.rpc.protocol.AbstractInvoker#doInvoke这个方法，DubboInvoker
